azavea/echo-locator
# Display BHA/Realtor.com listings

## Overview

This PR implements buttons to fetch and display BHA and Realtor.com listings as markers on the map. The buttons toggle between "show"/"hide" actions and the map markers should be appropriately visible/hidden when each button is clicked. Map marker interactivity will be handled separately, but button functionality should behave similar to [the demo site](https://demo.echosearch.org/map).


### Demo

![Screen Shot 2021-11-08 at 2 59 58 PM](https://user-images.githubusercontent.com/36374797/140810703-e56fc61e-f1b3-4fe2-95b4-a075c9a87081.png)
On Select Show BHA:
![Screen Shot 2021-11-08 at 3 00 38 PM](https://user-images.githubusercontent.com/36374797/140810739-4ce6a6f9-877b-41f7-9b44-033cf0da6a81.png)
On Select Show Realtor:
![Screen Shot 2021-11-08 at 3 00 56 PM](https://user-images.githubusercontent.com/36374797/140810793-0b074408-ca63-46c7-8145-3a568b82775f.png)

### Notes

PR-specific notes: 
1. Loading spinner, button colors, map marker colors and top bar styling were set to match demo site
2. The Realtor.com API response includes 100+ listings for a neighborhood. While the demo site displays all listings returned from the request, we may eventually include a limit or pagination.
3. The BHA API response does not include `lat`, `lon` values and the address must be forward geocoded. Mapbox API does have an endpoint for batch geocoding, however it involves reaching out a sales rep and paying extra. Right now, there are only a handful of BHA listings per neighborhood so making forward geocode requests individually isn't burdensome, but might be necessary with an increase in BHA listings.

Larger project-specific notes:
1. This project implements `redux-actions` which is no longer actively maintained. This PR continues to use `redux-actions` to handle state but it would be worth eventually looking into more stable redux utility packages.
2. This project handles async logic in action dispatches through custom middleware: `fetch`. By default, if no `error` parameter is present then, on `error`, `fetchError` is dispatched. We currently don't have any reducer to handle a `fetchError` dispatch and we have multiple action creators that depend on this action for error handling. This branch includes custom error handling, however we should create a separate issue that handles `fetchError` more broadly.


## Testing Instructions

 * Run `./scripts/server`
 * Log in and select a neighborhood (select 'Boston - Allston – 02134' neighborhood for BHA listings)
 * Confirm top bar and buttons appear above map
 * Select 'Show BHA' and confirm listings appear, select 'Hide BHA' and confirm listings disappear
 * Select 'Show Realtor' and confirm listings appear, select 'Hide Realtor' and confirm listings disappear
 * Select a new neighborhood and confirm listings clear
 * Simulate an API error (either by dropping a letter in the url or blocking request) and confirm that an error message appears next to buttons


Resolves #356
